javaslang/src/main/java/javaslang/collection/Tree.java

```diff
@@ -748,12 +771,12 @@ public boolean isEmpty() {
 
         @Override
         public boolean isLeaf() {
-            return children.isEmpty();
+            return size == 1;
         }
 
         @Override
         public int size() {
-            return 1 + children.foldLeft(0, (acc, child) -> acc + child.length());
+            return size;
```

Comment:
Hmmm... I don't see why a `Lazy` should be a real compromise, `get`-ing its value should be a lot faster than recalculating a tree/map/list size.

If it is, maybe we should change the double-check locking in `Lazy#get` (which many consider an anti-pattern anyway, especially since you're not checking against a `volatile`) to a static singleton.

I will experiment with it, let's see what the performance is of both :)